core/src/main/scala/org/apache/spark/deploy/history/HistoryPage.scala

```diff
@@ -88,4 +90,8 @@ private[history] class HistoryPage(parent: HistoryServer) extends WebUIPage("")
   private def makePageLink(showIncomplete: Boolean): String = {
     UIUtils.prependBaseUri("/?" + "showIncomplete=" + showIncomplete)
   }
+
+  private def isApplicationCompleted(appInfo: ApplicationInfo): Boolean = {
```

Comment:
The function `isApplicationCompleted` is defined 3 times in this patch.
I suggest adding a method `isCompleted` in class `ApplicationInfo`

Reply:
My first commit had this method in ApplicationInfo, called completed

@vanzin:
`+case class ApplicationInfo private[spark](
...
+    def completed: Boolean = {`
Is this used anywhere?

I answered we have it in HistoryPage (forgot to mention usage in the test)

@vanzin: Can we move this logic there instead? That avoids adding a new visible property in a public object.

I moved it, noting the code duplication in a comment on PR. 
The question is which is better: a bit of duplication (used 3 and defined 2 times, once in application and once in test code) or extending the public API? 
My gut feeling is that the small duplication is preferable in this case, but I'm a newbie in Spark development so I'm easy to convince that we have different preferences :)

Reply:
Duplication is mostly in tests so I don't care much, but if it really bothers someone, you can add this method in some utils object somewhere. It should just not be added to the public API.

Reply:
I've checked the status and the history packages and found no suitable object. 
As it is only one extra copy of a tiny method and it's in test code I don't think it would worth the extra complexity of introducing a new object, so I think it's best to leave it as it is. 
If we use it more frequently we can extract into a util object.